### PR TITLE
add status to the README

### PR DESCRIPTION
**Pull request by xmulligan**
Following #4 #37 and https://github.com/cilium/design-cfps/pull/9#issuecomment-2218325481

I've taken a stab at outlining a status for the CFP lifecycle. On this first pass I've tried to keep it pretty simple. Unmerged CFP is draft, merged is implementable, if a CFP is updated status is updated, and if it hasn't moved forward or was just an idea it is dormant.

Practically speaking that means that any CFP that has one committer's approval is now implementable. Hopefully, this will make it more clear that once there is alignment on the CFP, we can merge it so they aren't stuck as open PRs forever. We can then updated a CFP later as we see fit.

### CODE REVIEW

File: README.md
Comment (joestringer) [reply README.md:78]: Coming back to this I think this could be more confusing, it'd be clearer to keep dormant something like this. For instance if someone stumbles across a proposal and sees that it's merged with "draft", they wouldn't know if it's active or not. Dormant status makes this more clear.

Comment (joestringer) [inline README.md:78]:
```suggestion
5. Dormant: This status is given to a CFP that has not been
```

File: README.md
Comment (joestringer) [inline README.md:83]: nit: Wording changes just to simplify a bit.
```suggestion
currently a focus of the project. Any CFP where there is no active effort to
build the solution can have the dormant status. This serves as a way to preserve
previous discussions on a solution, but where the solution was not yet agreed
upon as implementable. Dormant CFPs can be reactivated
```

File: README.md
Comment (joestringer) [inline README.md:84]: Could perhaps add something to the effect that "The next step for a dormant CFP is to amend it to be Implementable."

**Review by joestringer — COMMENTED** [review]
Makes sense to me in principle. Couple of minor pieces of feedback:

- Updated status is a bit unclear to me - I think it's similar to Implementable but just demonstrates that the design was revised since the first design? But still may not be fully implemented.

- Might want an "implemented" status

- I briefly considered whether "dormant" should be separate from "draft", and figured that yes it probably makes sense to keep both separately, to help communicate where there is active interest on the CFP.

File: README.md
Comment (joestringer) [reply README.md:78]: I'm also imagining that we could have a dormant CFP that then later gets resurrected, so someone picks it up, makes proposed changes, re-opens the discussion, then when agreed upon, the dormant CFP moves to implementable.

File: README.md
Comment (xmulligan) [inline README.md:84]:
```suggestion
at any time if there is interest in the project. The next step for a dormant CFP is to amend it to be Implementable.
```

File: README.md
Comment (xmulligan) [reply README.md:78]: Yea exactly. I think leaving a paper trail that someone can pick up on later if they are interested.

File: README.md
Comment (xmulligan) [reply README.md]: Oh I like this idea, let me write up a suggestion

File: README.md
Comment (xmulligan) [inline README.md:76]:
```suggestion
4. Implemented, Release X.Y: Everything listed in the CFP as part of the proposal (not future milestones) is merged into the repository with the listed release.
```

File: README.md
Comment (xmulligan) [reply README.md]: added

File: README.md
Comment (joestringer) [reply README.md:76]: :ship:

File: README.md
Comment (joestringer) [inline README.md:74]: Not sure how I missed this before but the numbers in this list jump straight from 2 to 4, skipping 3.

**Comment by joestringer** [comment]
Oh I think Nick is actually OOO, so no need to block this PR on his response. I just figured he could potentially provide some good feedback.

Once you've revised the latest wording I can take a fresh look over the full text. We could also just advertise this a bit more broadly and set a deadline to merge. I think that after some minor editing this should be plenty good enough to start using and we can further amend the process from there as needed.

**Review by youngnick — COMMENTED** [review]
LGTM with @joestringer's changes. I think keeping it simple for now is a great idea, the Gateway API process has evolved over time, but YAGNI applies to processes as much as code. 😊

File: README.md
Comment (xmulligan) [inline README.md:78]:
```suggestion
4. Dormant: This status is given to a CFP that has not been
```

**Comment by xmulligan** [comment]
Sorry just catching up after vacation too. I've slightly changed the wording to released so we aren't going between implementable and implemented. I think it is ready for another read through and I can bring it to the developers meeting too.

File: README.md
Comment (bowei) [inline README.md:60]: Might want to reference related projects: https://github.com/kubernetes/enhancements/blob/master/keps/sig-architecture/0000-kep-process/README.md

Comment (bowei) [reply README.md:60]: If anything is relevant there, also ref: 

https://gateway-api.sigs.k8s.io/geps/overview/

along with rationale why the states exist.

Comment (xmulligan) [reply README.md:60]: Good point on your first comment, I've opened an issue for it here https://github.com/cilium/design-cfps/issues/47

File: README.md
Comment (xmulligan) [inline README.md:82]:
```suggestion
at any time if there is interest in the project. The next step for a dormant CFP is to amend it to be Implementable.

5. Declined: This proposal was considered by the community but ultimately rejected.
```

**Comment by xmulligan** [comment]
@bowei Thanks for pointing out the GEP process. I've looked through it and added `Declined` as a status too. I'll also add my thoughts on why I didn't add the other statuses below

`Completed` matches `Released`
`Standard` and `Experimental` match more to feature maturity/stability. So far in the Cilium project history, we haven't really tracked feature maturity. At a very high level there is the [Major Feature Status](https://docs.cilium.io/en/stable/community/roadmap/#major-feature-status), however we currently don't have clear guidelines around them and trying to add this right now to the CFP process would seem like a major change to the project workflow and I would try to avoid that right now unless we have a clear idea about what we are trying to get out of it
`Implementable`, `Provisional`,` `Prototyping` seem to roughly match `Implementable`. It is not quite clear to me the benefit of having these three different statuses to say we roughly agree on the idea, but haven't started all of the coding yet. Do you think we should have multiple statuses here?

**Comment by youngnick** [comment]
I think that it's better to start with less states, and add them if we need them later. I agree with adding `Declined`, but I think that what's there now is enough for a first iteration.

File: README.md
Comment (bowei) [inline README.md:46]: One of the reasons in other projects we merge proposals in progress and proposals that never make it is as a form of documentation on conversations on why/why the project did something.

I suggest we allow anything that has substantial conversations to be documented in a CFP (i.e. merged) -- even if in the end we decide not to go ahead.

File: README.md
Comment (bowei) [inline README.md:84]: Suggest add:

"The community may come back to the proposal in the future but a new CFP should be used. Rejected proposals are useful as documentation on why the given proposal did not make sense."

File: README.md
Comment (bowei) [inline README.md:65]: See my comment above -- it would be good to mandate that a Github issue be opened to track progress from Draft state i.e. make it a MUST that there is an issue. This makes it very easy for the project to track all draft proposals that pending (e.g. for triage).

File: README.md
Comment (bowei) [reply README.md:46]: (see my comment below with respect to always filing a feature).

**Comment by bowei** [comment]
Just so we don't forget, we may want to backpatch the existing CFPs with the status field for completeness...

File: README.md
Comment (joestringer) [reply README.md:46]: I'd like us to encourage this. :+1:

File: README.md
Comment (joestringer) [reply README.md:65]: Agree. I think we already have this expectation at some level, because the number of the CFP comes from the issue number on the corresponding repository. Even if it's a stub we should just make that part of the standard workflow.

Comment (joestringer) [reply README.md:65]: (Worth noting though that for any formal CFP there will be a PR in this repository, so tracking feels like it should be relatively straightforward by just listing the open PRs in this repository... but regardless of that I think this is worthwhile.)

File: README.md
Comment (joestringer) [inline README.md:76]: How much effort should we expect authors of dormant CFPs to put into clarifying the status of them?

One of the goals around "Dormant" status is I think to capture ideas that were considered but are not currently a priority. Sometimes these proposals may have aspects which are not agreed upon. While I can see the use of merging a record of the discussion into the repository, I worry a little that we might merge a dormant CFP with aspects that have no consensus, then subsequently someone may pick up the dormant PR and not address the concerns from those aspects with no consensus. The resulting PR might just look like a general tidy-up / improvement of a previous CFP.

Ultimately I guess this would just be up to the rigor of the review for when the status is moved to "Implementable", maybe the answer is that simple - make sure the full CFP has been re-reviewed prior to changing the status.

Comment (joestringer) [reply README.md:76]: @bowei this is related to [your thread above](https://github.com/cilium/design-cfps/pull/42#discussion_r1702120588), would be curious about your thoughts on this.

Comment (joestringer) [reply README.md:76]: It looks like at least https://github.com/cilium/design-cfps/pull/7 and https://github.com/cilium/design-cfps/pull/36 are in this category.

File: README.md
Comment (joestringer) [reply README.md:84]: > why the given proposal did not make sense.

How do we document this part?

If there's a PR and it's closed as not planned, I think the UX should be fairly clear. On the other hand, if the PR is merged into the repository, then the PR itself needs to be updated to encode this information. One of the things we've struggled with a bit as we start to establish the CFP process is that for the most part, developers don't want to document what they're designing. If you then decline their proposal and ask them to put additional effort into documenting why, then they're probably not all that motivated to comply. On the other hand, I guess as committers/maintainers we could take on that work, but in that case we also (a) need buy-in from those folks and (b) could benefit from an example workflow that describes what those folks should do in order to document the decision and why we feel it's important (helps with buy-in).

File: README.md
Comment (xmulligan) [inline README.md:45]:
```suggestion
For a CFP to be considered implementable, a Cilium committer needs to approve it.
```

File: README.md
Comment (xmulligan) [reply README.md:46]: I agree with that too, is there anything else you think we should add to the wording here?

File: README.md
Comment (xmulligan) [inline README.md:84]:
```suggestion
5. Declined: This proposal was considered by the community but ultimately rejected. The community may come back to the proposal in the future but a new CFP should be used. Rejected proposals can be useful as documentation on why the given proposal did not make sense.
```

Comment (xmulligan) [reply README.md:84]: I've updated the wording to "can be" and leaving the implementation that Joe is discussing above to be worked out when we come to that bridge.

File: README.md
Comment (xmulligan) [inline README.md:65]:
```suggestion
1. Draft: A CFP in any form (Google doc, HackMD, PR, ect.) with associated Github issue that has not yet
```

Comment (xmulligan) [reply README.md:65]: updated the wording slightly to make Github issue MUST

File: README.md
Comment (joestringer) [reply README.md]: I think it's worth noting that the default behaviour if we don't resolve that question will be either (a) to leave PRs that have been declined open, or to (b) just close them without merging. Either of these options takes minimal to no effort vs. figuring out the standard we expect in order to merge a declined CFP.

That said, I'm OK with this outcome and we can always improve later, I just want to set expectations that even if we agree in principle that merging declined CFPs is OK, unless we have a specific pattern to follow, we will *not* merge declined PRs.

File: README.md
Comment (bowei) [reply README.md:76]: Yeah, I think this is a responsibility for the flip to "implementable" -- don't just do it without clear signal the proposal is agreed upon. This is something that would be easy to resolve without added procedure -- it should be hard to have a submarine proposal and anything that wasn't supposed to be merged would probably be backed out at some point with the biggest loss being work...

Comment (xmulligan) [reply README.md:76]: Agree, I think that is part of the switch to implementable and just requires people to be thoughtful in their reviews

File: README.md
Comment (xmulligan) [inline README.md:48]:
```suggestion
100% chance the code will be accepted. Dormant and declined CFPs are
also an important part of documentation for the project. CFPs with theses
statuses can also be merged with approval from one committer.
```

File: README.md
Comment (xmulligan) [reply README.md:46]: added some suggested wording about dormant and declined also being important and how they can be merged, let me know what you think

File: README.md
Comment (joestringer) [reply README.md:48]: ^ Re: last suggestion, first line already mentions committer approval - no need to mention it twice.

File: README.md
Comment (joestringer) [inline README.md:62]:
```suggestion
The status of a CFP indicates its maturity. There are five different statuses
```

**Review by joestringer — APPROVED** [review]
Pretty good, couple of minor bits of feedback on wordsmithing.

File: README.md
Comment (xmulligan) [inline README.md:48]:
```suggestion
A CFP can be merged with the approval of one committer. A CFP merged as implementable means the proposal is viable to be implemented, but not that there is a 100% chance the code will be accepted. Dormant and declined CFPs are also an important part of documentation for the project and should also be merged.
```

Comment (xmulligan) [reply README.md:48]: ok, slightly changed around the wording to reflect that

**Comment by bowei** [comment]
LGTM

**Review by bowei — APPROVED** [review]

**Pull request merged**
